### PR TITLE
#166457277 change email verification implementation

### PR DESCRIPTION
**Pull request by encodedBicoding**
### What does this PR do?
Changes the implementation used on the feature to verify email.
##### Why was it changed?
The former code, was vulnerable, due to the fact that other token generated via signup and signin could be used to verify a user. 
this new code checks for the token supplied against the token in the users `email_confirm_code` section
#### Description of Task to be completed?
Change Logic implementation
#### How should this be manually tested?
N/A
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166457277](https://www.pivotaltracker.com/story/show/166457277)
#### Screenshots (if appropriate)
Screenshot (59)

#### Questions:
N/A

### CODE REVIEW

**Comment by emp-daisy** [comment]
Can you look into the comments raised by HOUND?

File: server/routes/userRoute.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ const userRoute = express.Router();
 
 userRoute.post('/signup', Validate.signup, UserController.signUp); 
 userRoute.post('/signin', Validate.signin, UserController.emailSignin);
-userRoute.get('/verifyEmail', UserController.verifyEmailLink);
+userRoute.get('/verifyemail', UserController.verifyEmailLink);
```
Comment (emp-daisy) [inline server/routes/userRoute.js:9]: Update the docs to reflect this change

File: server/controllers/userController.js [hunk 2]
```diff
@@ -116,15 +116,15 @@ class UserController {
...
     try {
-      const { token } = req.query;
-      const payload = auth.verifyMailToken(token);
-
-      if (!payload) return util.errorStatus(res, 400, 'Invalid Verification Link');
-
-      const { email } = payload;
-
+      const { token, id } = req.query;
+      const user = await models.Users.findByPk(id);
```
Comment (emp-daisy) [inline server/controllers/userController.js:140]: What is "findByPk"?

Comment (chuxmykel) [reply server/controllers/userController.js:140]: Sequelize is deprecating findById in v5 I think and are replacing it with findByPk(). PK = Primary Key.

Comment (emp-daisy) [reply server/controllers/userController.js:140]: Oh, I see. I wasn't aware of that.

File: server/controllers/userController.js [hunk 3]
```diff
@@ -116,15 +116,15 @@ class UserController {
...
   static async verifyEmailLink(req, res) {
     try {
-      const { token } = req.query;
-      const payload = auth.verifyMailToken(token);
-
-      if (!payload) return util.errorStatus(res, 400, 'Invalid Verification Link');
-
-      const { email } = payload;
-
+      const { token, id } = req.query;
```
Comment (emp-daisy) [inline server/controllers/userController.js:122]: Before querying the DB, check the token and id value are defined else return Invalid link or something more descriptive

File: server/controllers/userController.js [hunk 4]
```diff
@@ -116,15 +116,15 @@ class UserController {
...
+      const user = await models.Users.findByPk(id);
+      if(!user) return util.errorStatus(res, 401, 'Not authorized');
```
Comment (emp-daisy) [inline server/controllers/userController.js]: This should also be an Invalid link not unauthorized

File: server/controllers/userController.js (hunk 3 above)
Comment (encodedBicoding) [reply server/controllers/userController.js:122]: Ok noted.

**Review by chuxmykel — APPROVED** [review]
LGTM

**Review by PelumiAlesh — APPROVED** [review]

File: server/__test__/user.spec.js [hunk 5]
```diff
@@ -349,12 +350,57 @@ describe('test for verifying email', () => {
       expect(res.body).toMatchSnapshot();
     })
   });
-  it('Should fail if token is invalid', async done => {
+  it('Should fail if token doesn\'t match the user email_confirm_code token', async done => {
      server()
-    .get(`${url}/auth/verifyEmail?token=jkjidsjijesnjnudsufjfjewjisdjijsdfnjsifjisdfjsidjfisdjfi`)
+    .get(`${url}/auth/verifyemail?token=${token.substring(2, 30)}&id=${userId}`)
```
Comment (ocranbillions) [inline server/__test__/user.spec.js:355]: lol, nice logic! `token.substring(2, 30)`

**Pull request merged**